staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -596,21 +609,29 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 			growSlice(v, 2048, len(getResp.Kvs))
 		}
 
-		// take items from the response until the bucket is full, filtering as we go
-		for _, kv := range getResp.Kvs {
-			if paging && int64(v.Len()) >= pred.Limit {
-				hasMore = true
-				break
-			}
-			lastKey = kv.Key
+		chunks := splitKVsIntoChunks(s.listTransformationConcurrency, getResp.Kvs)
+		fanOutChannels := make([]<-chan *transformedKVPair, len(chunks))
+		for i, c := range chunks {
+			fanOutChannels[i] = s.transformChunk(ctx, c)
+		}
 
-			data, _, err := s.transformer.TransformFromStorage(kv.Value, authenticatedDataString(kv.Key))
-			if err != nil {
-				return storage.NewInternalErrorf("unable to transform key %q: %v", kv.Key, err)
-			}
+		// To preserve the order of items in the response, we iterate over the channels, as opposed to
+		// creating a single multiplexed channel and accepting transformed items as they arrive.
+	ChannelLoop:
+		for _, c := range fanOutChannels {
+			for transformed := range c {
+				if paging && int64(v.Len()) >= pred.Limit {
```

Comment:
As currently written:
Yes, we will transform all items, and some of such items may not be returned (see below).
This could happen during paging and when the storage predicate contains both: **filter** and **limit**. Since we can't apply the filter until all items are transformed, we may go over the limit and discard some transformed items. This is somewhat analogous to speculative processing.

I believe I should be able to address this scenario. I will work on a fix.